SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracer.cpp

```diff
@@ -0,0 +1,322 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "Interop/Connection/SpatialEventTracer.h"
+
+#include "UObject/Class.h"
+#include "GameFramework/Actor.h"
+
+#include <WorkerSDK/improbable/c_trace.h>
+
+DEFINE_LOG_CATEGORY(LogSpatialEventTracer);
+
+using namespace SpatialGDK;
+using namespace worker::c;
+
+void MyTraceCallback(void* UserData, const Trace_Item* Item)
+{
+	switch (Item->item_type)
+	{
+	case TRACE_ITEM_TYPE_EVENT:
+	{
+		const Trace_Event& Event = Item->item.event;
+
+		// TODO: remove temporary filtering?
+		if (Event.type == FString("network.receive_raw_message") ||
+			Event.type == FString("network.receive_udp_datagram") ||
+			Event.type == FString("network.send_raw_message") ||
+			Event.type == FString("network.send_udp_datagram") ||
+			Event.type == FString("worker.dequeue_op") ||
+			Event.type == FString("worker.enqueue_op"))
+		{
+			return;
+		}
+
+		unsigned long int span1 = *reinterpret_cast<const unsigned long int*>(&Event.span_id.data[0]);
+		unsigned long int span2 = *reinterpret_cast<const unsigned long int*>(&Event.span_id.data[8]);
+		UE_LOG(LogSpatialEventTracer, Warning, TEXT("Span: %lu%lu, Type: %s, Message: %s, Timestamp: %lu"),
+			span1, span2, *FString(Event.type), *FString(Event.message), Event.unix_timestamp_millis);
+
+		if (Event.data != nullptr)
+		{
+			uint32_t DataFieldCount = Trace_EventData_GetFieldCount(Event.data);
+
+			TArray<const char*> Keys;
+			Keys.SetNumUninitialized(DataFieldCount);
+			TArray<const char*> Values;
+			Values.SetNumUninitialized(DataFieldCount);
+
+			Trace_EventData_GetStringFields(Event.data, Keys.GetData(), Values.GetData());
```

Comment:
Just a reminder that we should ask for an API for binary encoding from worker! 

We could use Unreal serialization to pass through and automatically handle serialization with FMemoryArchive, just an idea!

Reply:
I take this back, it's not in scope and neither does it align with the distributed debugging :)